### PR TITLE
WIP: Add render

### PR DESCRIPTION
**Pull request by hexfusion**
This PR is still very green and I will leave the verbose commits until its time to rebase for a final PR against upstream.

### functionality
- [x] Populate etcd-member-pod.yaml template. @jeckersb https://github.com/hexfusion/cluster-etcd-operator/issues/3
- [x] Install etcd-conf on bootstrap to /etc/etcd/etcd.conf @hexfusion https://github.com/hexfusion/cluster-etcd-operator/issues/6
- [x] Install etcd CAs on bootstrap where it is expected by static pod @hexfusion 
- [x] fix existing tests @jeckersb
- [x] fix travis @hexfusion 
- [x] Install namespaces, services, secrets and configmaps to support etcd static pod for bootstrap node. @hexfusion 

### testing
For now I am testing this in a pretty basic way but its all I have and open to input/improvements. Moving forward we can move the bash below into bootkube and use the custom installer to test.

First I build and tag docker.

```
docker build -f Dockerfile . -t quay.io/hexfusion/cluster-etcd-operator:$tag
docker push quay.io/hexfusion/cluster-etcd-operator:$tag
```

Now on a bootstrap node 

```
#!/usr/bin/env bash

set -e

mkdir --parents /etc/kubernetes/{manifests,bootstrap-configs,bootstrap-manifests} /etc/etcd

if ! podman inspect registry.svc.ci.openshift.org/origin/release:4.2 &>/dev/null; then
    echo "Pulling release image..."
    podman pull --quiet registry.svc.ci.openshift.org/origin/release:4.2
fi

# convert the release image pull spec to an "absolute" form if a digest is available - this is
# safe to resolve after the actions above because podman will not pull the image once it is
# locally available
if ! release=$( podman inspect registry.svc.ci.openshift.org/origin/release:4.2 -f '{{ index .RepoDigests 0 }}' ) || [[ -z "${release}" ]]; then
        echo "Warning: Could not resolve release image to pull by digest" 2>&1
        release="registry.svc.ci.openshift.org/origin/release:4.2"
fi

ETCD_IMAGE=$(podman run --quiet --rm ${release} image etcd)
KUBE_CLIENT_AGENT_IMAGE=$(podman run --quiet --rm ${release} image kube-client-agent)
MACHINE_CONFIG_SETUP_ETCD_ENV_IMAGE=$(podman run --quiet --rm ${release} image setup-etcd-environment)

cd /opt/openshift

ETCD_OPERATOR_IMAGE=quay.io/hexfusion/cluster-etcd-operator:latest

    # shellcheck disable=SC2154
    podman run \
    --quiet \
    --volume "$PWD:/assets:z" \
    --volume /etc/kubernetes:/etc/kubernetes:z \
    --volume /etc/etcd:/etc/etcd:z \
    "${ETCD_OPERATOR_IMAGE}" \
    /usr/bin/cluster-etcd-operator render \
    --etcd-ca=/assets/tls/etcd-ca-bundle.crt \
    --etcd-metric-ca=/assets/tls/etcd-metric-ca-bundle.crt \
    --manifest-etcd-image "${ETCD_IMAGE}" \
    --etcd-discovery-domain "sbatsche.hexfusion.local" \
    --manifest-setup-etcd-env-image "${MACHINE_CONFIG_SETUP_ETCD_ENV_IMAGE}" \
    --manifest-image-pull-policy "TEST" \
    --manifest-kube-client-agent-image "${KUBE_CLIENT_AGENT_IMAGE}" \
    --asset-input-dir ./assets/tls \
    --asset-output-dir ./assets/etcd-bootstrap \
    --config-output-file ./assets/etcd-bootstrap/config
```

### CODE REVIEW

File: bindata/bootkube/bootstrap-manifests/etcd-member-pod.yaml [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+            --ipaddrs=${ETCD_IPV4_ADDRESS} \
+
+      [ -e /etc/ssl/etcd/system:etcd-metric:${ETCD_DNS_NAME}.crt -a \
+        -e /etc/ssl/etcd/system:etcd-metric:${ETCD_DNS_NAME}.key ] || \
+        kube-client-agent \
+          request \
+            --kubeconfig=/etc/kubernetes/kubeconfig \
+            --orgname=system:etcd-metrics \
+            --assetsdir=/etc/ssl/etcd \
+            --dnsnames=localhost,etcd.kube-system.svc,etcd.kube-system.svc.cluster.local,etcd.openshift-etcd.svc,etcd.openshift-etcd.svc.cluster.local,${ETCD_WILDCARD_DNS_NAME} \
```
Comment (hexfusion) [inline bindata/bootkube/bootstrap-manifests/etcd-member-pod.yaml]: we need to move this logic to render from MCO, meaning replace hard coded values with {{etcdPeerCertDNSNames}} and {{etcdServerCertDNSNames}} and populate in render.go

File: pkg/cmd/render/render.go [hunk 2]
```diff
@@ -100,88 +126,169 @@ func (r *renderOpts) Complete() error {
...
+
 	if err := r.manifest.ApplyTo(&renderConfig.ManifestConfig); err != nil {
 		return err
 	}
 	if err := r.generic.ApplyTo(
 		&renderConfig.FileConfig,
-		genericrenderoptions.Template{FileName: "defaultconfig.yaml", Content: v311_00_assets.MustAsset(filepath.Join(bootstrapVersion, "etcd", "defaultconfig.yaml"))},
+		options.Template{FileName: "defaultconfig.yaml", Content: v420_00_assets.MustAsset(filepath.Join(bootstrapVersion, "etcd", "defaultconfig.yaml"))},
 		mustReadTemplateFile(filepath.Join(r.generic.TemplatesDir, "config", "bootstrap-config-overrides.yaml")),
 		mustReadTemplateFile(filepath.Join(r.generic.TemplatesDir, "config", "config-overrides.yaml")),
```
Comment (hexfusion) [inline pkg/cmd/render/render.go:209]: we aren't using most/all of these we need to clean this up.

**Comment by beekhof** [comment]
/lgtm 

While there may be improvements to be found, I would be in favour of merging this sooner rather than later.

**Comment by hexfusion** [comment]
@ I am going to reduce my work down to a few commits and merge.

**Comment by hexfusion** [comment]
@beekhof I am going to reduce my work down to a few commits and merge.

**Pull request merged**